diff --git a/init/do_mounts.c b/init/do_mounts.c
--- a/init/do_mounts.c
+++ b/init/do_mounts.c
@@ -503,8 +503,27 @@ void __init change_floppy(char *fmt, ...)
 }
 #endif
 
+#ifdef CONFIG_AXFS_FIRST_MOUNT_AXFS_AS_ROOTFS
+static int __init mount_axfs_root(void)
+{
+	create_dev("/dev/root", ROOT_DEV);
+	if (root_mount_data &&
+	    do_mount_root("/dev/root", "axfs", root_mountflags,
+					root_mount_data) == 0)
+		return 1;
+	return 0;
+}
+#endif
+
 void __init mount_root(void)
 {
+#ifdef CONFIG_AXFS_FIRST_MOUNT_AXFS_AS_ROOTFS
+	if (MAJOR(ROOT_DEV) == UNNAMED_MAJOR) {
+		if (mount_axfs_root())
+			return;
+		printk(KERN_ERR "VFS: Unable to mount axfs root.\n");
+	}
+#endif
 #ifdef CONFIG_ROOT_NFS
 	if (ROOT_DEV == Root_NFS) {
 		if (mount_nfs_root())
