diff -r ede538c7c673 arch/um/include/kern_util.h
--- a/arch/um/include/kern_util.h	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/include/kern_util.h	Fri Nov 21 03:51:29 2008 -0800
@@ -6,9 +6,9 @@
 #ifndef __KERN_UTIL_H__
 #define __KERN_UTIL_H__
 
-#include "linux/threads.h"
 #include "sysdep/ptrace.h"
 #include "sysdep/faultinfo.h"
+#define NR_CPUS 1
 
 extern int ncpus;
 extern char *linux_prog;
diff -r ede538c7c673 arch/um/include/sysdep-i386/stub.h
--- a/arch/um/include/sysdep-i386/stub.h	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/include/sysdep-i386/stub.h	Fri Nov 21 03:51:29 2008 -0800
@@ -12,6 +12,7 @@
 extern void stub_segv_handler(int sig);
 extern void stub_clone_handler(void);
 
+#define PAGE_SHIFT 12
 #define STUB_SYSCALL_RET EAX
 #define STUB_MMAP_NR __NR_mmap2
 #define MMAP_OFFSET(o) ((o) >> PAGE_SHIFT)
diff -r ede538c7c673 arch/um/kernel/main.c
--- a/arch/um/kernel/main.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/kernel/main.c	Fri Nov 21 03:51:29 2008 -0800
@@ -12,7 +12,7 @@
 #include <sys/resource.h>
 #include <sys/mman.h>
 #include <sys/user.h>
-#include <asm/page.h>
+#include "mode-tt.h"
 #include "user_util.h"
 #include "kern_util.h"
 #include "mem_user.h"
diff -r ede538c7c673 arch/um/kernel/skas/clone.c
--- a/arch/um/kernel/skas/clone.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/kernel/skas/clone.c	Fri Nov 21 03:51:29 2008 -0800
@@ -3,7 +3,7 @@
 #include <sys/mman.h>
 #include <sys/time.h>
 #include <asm/unistd.h>
-#include <asm/page.h>
+#include <sys/user.h>
 #include "ptrace_user.h"
 #include "skas.h"
 #include "stub-data.h"
diff -r ede538c7c673 arch/um/kernel/skas/mem_user.c
--- a/arch/um/kernel/skas/mem_user.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/kernel/skas/mem_user.c	Fri Nov 21 03:51:29 2008 -0800
@@ -8,7 +8,6 @@
 #include <string.h>
 #include <sys/mman.h>
 #include <sys/wait.h>
-#include <asm/page.h>
 #include <asm/unistd.h>
 #include "mem_user.h"
 #include "mem.h"
@@ -24,6 +23,10 @@
 #include "uml-config.h"
 #include "sysdep/ptrace.h"
 #include "sysdep/stub.h"
+
+#define PAGE_SHIFT	12
+#define PAGE_SIZE	(1UL << PAGE_SHIFT)
+#define PAGE_MASK	(~(PAGE_SIZE-1))
 
 extern unsigned long batch_syscall_stub, __syscall_stub_start;
 
diff -r ede538c7c673 arch/um/kernel/trap_user.c
--- a/arch/um/kernel/trap_user.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/kernel/trap_user.c	Fri Nov 21 03:51:29 2008 -0800
@@ -9,7 +9,6 @@
 #include <signal.h>
 #include <sys/time.h>
 #include <sys/wait.h>
-#include <asm/page.h>
 #include <asm/unistd.h>
 #include <asm/ptrace.h>
 #include "init.h"
diff -r ede538c7c673 arch/um/os-Linux/aio.c
--- a/arch/um/os-Linux/aio.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/os-Linux/aio.c	Fri Nov 21 03:51:29 2008 -0800
@@ -80,8 +80,7 @@
                                 .aio_nbytes	= len,
                                 .aio_offset	= offset,
                                 .aio_reserved1	= 0,
-                                .aio_reserved2	= 0,
-                                .aio_reserved3	= 0 });
+                                .aio_reserved2	= 0 });
 
         switch(type){
         case AIO_READ:
@@ -335,7 +334,7 @@
                 if(!aio_24){
                         printk("Disabling 2.6 AIO in tt mode\n");
                         aio_24 = 1;
-                } }), (void) 0);
+                } }), 0);
 
         if(!aio_24){
                 err = init_aio_26();
diff -r ede538c7c673 arch/um/os-Linux/process.c
--- a/arch/um/os-Linux/process.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/os-Linux/process.c	Fri Nov 21 03:51:29 2008 -0800
@@ -123,11 +123,9 @@
  * syscalls, and also breaks with clone(), which does not unshare the TLS.
  */
 
-inline _syscall0(pid_t, getpid)
-
 int os_getpid(void)
 {
-	return(getpid());
+  return(syscall(__NR_getpid));
 }
 
 int os_getpgrp(void)
diff -r ede538c7c673 arch/um/os-Linux/start_up.c
--- a/arch/um/os-Linux/start_up.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/os-Linux/start_up.c	Fri Nov 21 03:51:29 2008 -0800
@@ -18,7 +18,7 @@
 #include <sys/wait.h>
 #include <sys/mman.h>
 #include <asm/unistd.h>
-#include <asm/page.h>
+#include <sys/user.h>
 #include <sys/types.h>
 #include "user_util.h"
 #include "kern_util.h"
diff -r ede538c7c673 arch/um/os-Linux/sys-i386/registers.c
--- a/arch/um/os-Linux/sys-i386/registers.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/os-Linux/sys-i386/registers.c	Fri Nov 21 03:51:29 2008 -0800
@@ -127,6 +127,12 @@
 	memcpy(regs, exec_regs, HOST_FRAME_SIZE * sizeof(unsigned long));
 }
 
+#ifndef JB_PC
+#define JB_PC 5
+#define JB_SP 4
+#define JB_BP 3
+#endif
+
 void get_thread_regs(union uml_pt_regs *uml_regs, void *buffer)
 {
 	struct __jmp_buf_tag *jmpbuf = buffer;
diff -r ede538c7c673 arch/um/os-Linux/sys-x86_64/registers.c
--- a/arch/um/os-Linux/sys-x86_64/registers.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/os-Linux/sys-x86_64/registers.c	Fri Nov 21 03:51:29 2008 -0800
@@ -75,6 +75,12 @@
 	memcpy(regs, exec_regs, HOST_FRAME_SIZE * sizeof(unsigned long));
 }
 
+#ifndef JB_PC
+#define JB_PC 7
+#define JB_RSP 6
+#define JB_RBP 1
+#endif
+
 void get_thread_regs(union uml_pt_regs *uml_regs, void *buffer)
 {
 	struct __jmp_buf_tag *jmpbuf = buffer;
diff -r ede538c7c673 arch/um/os-Linux/tt.c
--- a/arch/um/os-Linux/tt.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/os-Linux/tt.c	Fri Nov 21 03:51:29 2008 -0800
@@ -18,7 +18,7 @@
 #include <sys/mman.h>
 #include <asm/ptrace.h>
 #include <asm/unistd.h>
-#include <asm/page.h>
+#include <sys/user.h>
 #include "user_util.h"
 #include "kern_util.h"
 #include "user.h"
diff -r ede538c7c673 arch/um/sys-i386/ptrace_user.c
--- a/arch/um/sys-i386/ptrace_user.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/sys-i386/ptrace_user.c	Fri Nov 21 03:51:29 2008 -0800
@@ -6,10 +6,10 @@
 #include <stdio.h>
 #include <errno.h>
 #include <unistd.h>
-#include <linux/stddef.h>
+#include <stddef.h>
 #include "ptrace_user.h"
 /* Grr, asm/user.h includes asm/ptrace.h, so has to follow ptrace_user.h */
-#include <asm/user.h>
+#include <sys/user.h>
 #include "kern_util.h"
 #include "sysdep/thread.h"
 #include "user.h"
diff -r ede538c7c673 arch/um/sys-i386/user-offsets.c
--- a/arch/um/sys-i386/user-offsets.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/arch/um/sys-i386/user-offsets.c	Fri Nov 21 03:51:29 2008 -0800
@@ -1,7 +1,8 @@
 #include <stdio.h>
 #include <signal.h>
+#include <stddef.h>
 #include <asm/ptrace.h>
-#include <asm/user.h>
+#include <sys/user.h>
 #include <linux/stddef.h>
 
 #define DEFINE(sym, val) \
@@ -47,8 +48,8 @@
 	OFFSET(HOST_SC_FXSR_ENV, _fpstate, _fxsr_env);
 
 	DEFINE(HOST_FRAME_SIZE, FRAME_SIZE);
-	DEFINE_LONGS(HOST_FP_SIZE, sizeof(struct user_i387_struct));
-	DEFINE_LONGS(HOST_XFP_SIZE, sizeof(struct user_fxsr_struct));
+	DEFINE_LONGS(HOST_FP_SIZE, sizeof(struct user_fpregs_struct));
+	DEFINE_LONGS(HOST_XFP_SIZE, sizeof(struct user_fpxregs_struct));
 
 	DEFINE(HOST_IP, EIP);
 	DEFINE(HOST_SP, UESP);
diff -r ede538c7c673 scripts/mod/sumversion.c
--- a/scripts/mod/sumversion.c	Thu Sep 18 13:17:45 2008 -0700
+++ b/scripts/mod/sumversion.c	Fri Nov 21 03:51:29 2008 -0800
@@ -7,6 +7,7 @@
 #include <ctype.h>
 #include <errno.h>
 #include <string.h>
+#include <limits.h>
 #include "modpost.h"
 
 /*
