diff --git a/include/asm-i386/linkage.h b/include/asm-i386/linkage.h
index f4a6eba..49863c3 100644
--- a/include/asm-i386/linkage.h
+++ b/include/asm-i386/linkage.h
@@ -2,8 +2,8 @@
 #define __ASM_LINKAGE_H
 
 #define asmlinkage CPP_ASMLINKAGE __attribute__((regparm(0)))
-#define FASTCALL(x)	x __attribute__((regparm(3)))
-#define fastcall	__attribute__((regparm(3)))
+#define FASTCALL(x)	x
+#define fastcall
 
 #define prevent_tail_call(ret) __asm__ ("" : "=r" (ret) : "0" (ret))
 
