diff -r 907888e95d29 drivers/mtd/chips/cfi_cmdset_0001.c
--- a/drivers/mtd/chips/cfi_cmdset_0001.c	Tue Sep 02 14:27:46 2008 -0700
+++ b/drivers/mtd/chips/cfi_cmdset_0001.c	Tue Sep 02 14:29:14 2008 -0700
@@ -67,9 +67,8 @@ static int cfi_intelext_partition_fixup(
 static int cfi_intelext_partition_fixup(struct mtd_info *, struct cfi_private **);
 
 static int cfi_intelext_point (struct mtd_info *mtd, loff_t from, size_t len,
-		     size_t *retlen, u_char **mtdbuf);
-static void cfi_intelext_unpoint (struct mtd_info *mtd, u_char *addr, loff_t from,
-			size_t len);
+		     size_t *retlen, void **virt, unsigned long *phys);
+static void cfi_intelext_unpoint(struct mtd_info *mtd, loff_t from, size_t len);
 
 static int get_chip(struct map_info *map, struct flchip *chip, unsigned long adr, int mode);
 static void put_chip(struct map_info *map, struct flchip *chip, unsigned long adr);
@@ -1018,7 +1017,8 @@ static int do_point_onechip (struct map_
 	return ret;
 }
 
-static int cfi_intelext_point (struct mtd_info *mtd, loff_t from, size_t len, size_t *retlen, u_char **mtdbuf)
+static int cfi_intelext_point(struct mtd_info *mtd, loff_t from, size_t len,
+		size_t *retlen, void **virt, unsigned long *phys)
 {
 	struct map_info *map = mtd->priv;
 	struct cfi_private *cfi = map->fldrv_priv;
@@ -1029,14 +1029,16 @@ static int cfi_intelext_point (struct mt
 	if (!map->virt || (from + len > mtd->size))
 		return -EINVAL;
 	
-	*mtdbuf = (void *)map->virt + from;
-	*retlen = 0;
-
 	/* Now lock the chip(s) to POINT state */
 
 	/* ofs: offset within the first chip that the first read should start */
 	chipnum = (from >> cfi->chipshift);
 	ofs = from - (chipnum << cfi->chipshift);
+
+	*virt = map->virt + cfi->chips[chipnum].start + ofs;
+	*retlen = 0;
+	if (phys)
+		*phys = map->phys + cfi->chips[chipnum].start + ofs;
 
 	while (len) {
 		unsigned long thislen;
@@ -1062,7 +1064,7 @@ static int cfi_intelext_point (struct mt
 	return 0;
 }
 
-static void cfi_intelext_unpoint (struct mtd_info *mtd, u_char *addr, loff_t from, size_t len)
+static void cfi_intelext_unpoint(struct mtd_info *mtd, loff_t from, size_t len)
 {
 	struct map_info *map = mtd->priv;
 	struct cfi_private *cfi = map->fldrv_priv;
diff -r 907888e95d29 drivers/mtd/devices/mtdram.c
--- a/drivers/mtd/devices/mtdram.c	Tue Sep 02 14:27:46 2008 -0700
+++ b/drivers/mtd/devices/mtdram.c	Tue Sep 02 14:29:14 2008 -0700
@@ -62,18 +62,22 @@ ram_erase(struct mtd_info *mtd, struct e
   return 0;
 }
 
-static int ram_point (struct mtd_info *mtd, loff_t from, size_t len, size_t *retlen, u_char **mtdbuf)
+static int ram_point(struct mtd_info *mtd, loff_t from, size_t len,
+		     size_t *retlen, void **virt, unsigned long *phys)
 {
   if (from + len > mtd->size)
     return -EINVAL;
-	
-  *mtdbuf = mtd->priv + from;
+
+	/* can we return a physical address with this driver? */
+	if (phys)
+		return -EINVAL;
+
+	*virt = mtd->priv + from;
   *retlen = len;
   return 0;
 }
 
-static void ram_unpoint (struct mtd_info *mtd, u_char *addr, loff_t from,
-			 size_t len)
+static void ram_unpoint(struct mtd_info *mtd, loff_t from, size_t len)
 {
   DEBUG(MTD_DEBUG_LEVEL2, "ram_unpoint\n");
 }
diff -r 907888e95d29 drivers/mtd/devices/phram.c
--- a/drivers/mtd/devices/phram.c	Tue Sep 02 14:27:46 2008 -0700
+++ b/drivers/mtd/devices/phram.c	Tue Sep 02 14:29:14 2008 -0700
@@ -59,19 +59,23 @@ static int phram_erase(struct mtd_info *
 }
 
 static int phram_point(struct mtd_info *mtd, loff_t from, size_t len,
-		size_t *retlen, u_char **mtdbuf)
+		       size_t *retlen, void **virt, unsigned long *phys)
 {
 	u_char *start = mtd->priv;
 
 	if (from + len > mtd->size)
 		return -EINVAL;
-	
-	*mtdbuf = start + from;
+
+	/* can we return a physical address with this driver? */
+	if (phys)
+		return -EINVAL;
+
+	*virt = mtd->priv + from;
 	*retlen = len;
 	return 0;
 }
 
-static void phram_unpoint(struct mtd_info *mtd, u_char *addr, loff_t from, size_t len)
+static void phram_unpoint(struct mtd_info *mtd, loff_t from, size_t len)
 {
 }
 
diff -r 907888e95d29 drivers/mtd/devices/pmc551.c
--- a/drivers/mtd/devices/pmc551.c	Tue Sep 02 14:27:46 2008 -0700
+++ b/drivers/mtd/devices/pmc551.c	Tue Sep 02 14:29:14 2008 -0700
@@ -139,7 +139,8 @@ static int pmc551_erase (struct mtd_info
         eoff_lo = end & (priv->asize - 1);
         soff_lo = instr->addr & (priv->asize - 1);
 
-	pmc551_point (mtd, instr->addr, instr->len, &retlen, &ptr);
+	pmc551_point(mtd, instr->addr, instr->len, &retlen,
+	              (void **)&ptr, NULL);
 
         if ( soff_hi == eoff_hi || mtd->size == priv->asize) {
                 /* The whole thing fits within one access, so just one shot
@@ -158,7 +159,8 @@ static int pmc551_erase (struct mtd_info
                         }
                         soff_hi += priv->asize;
 			pmc551_point (mtd,(priv->base_map0|soff_hi),
-				      priv->asize, &retlen, &ptr);
+				      priv->asize, &retlen,
+				      (void **)&ptr, NULL);
                 }
                 memset (ptr, 0xff, eoff_lo);
         }
@@ -174,7 +176,8 @@ out:
 }
 
 
-static int pmc551_point (struct mtd_info *mtd, loff_t from, size_t len, size_t *retlen, u_char **mtdbuf)
+static int pmc551_point(struct mtd_info *mtd, loff_t from, size_t len,
+			 size_t *retlen, void **virt, unsigned long *phys)
 {
         struct mypriv *priv = mtd->priv;
         u32 soff_hi;
@@ -191,6 +194,10 @@ static int pmc551_point (struct mtd_info
 		return -EINVAL;
 	}
 
+	/* can we return a physical address with this driver? */
+	if (phys)
+		return -EINVAL;
+
         soff_hi = from & ~(priv->asize - 1);
         soff_lo = from & (priv->asize - 1);
 
@@ -201,13 +208,13 @@ static int pmc551_point (struct mtd_info
 		priv->curr_map0 = soff_hi;
 	}
 
-	*mtdbuf = priv->start + soff_lo;
+	*virt = priv->start + soff_lo;
 	*retlen = len;
 	return 0;
 }
 
 
-static void pmc551_unpoint (struct mtd_info *mtd, u_char *addr, loff_t from, size_t len)
+static void pmc551_unpoint(struct mtd_info *mtd, loff_t from, size_t len)
 {
 #ifdef CONFIG_MTD_PMC551_DEBUG
 	printk(KERN_DEBUG "pmc551_unpoint()\n");
@@ -243,7 +250,7 @@ static int pmc551_read (struct mtd_info 
         soff_lo = from & (priv->asize - 1);
         eoff_lo = end & (priv->asize - 1);
 
-	pmc551_point (mtd, from, len, retlen, &ptr);
+	pmc551_point(mtd, from, len, retlen, (void **)&ptr, NULL);
 
         if (soff_hi == eoff_hi) {
                 /* The whole thing fits within one access, so just one shot
@@ -263,7 +270,8 @@ static int pmc551_read (struct mtd_info 
                                 goto out;
                         }
                         soff_hi += priv->asize;
-			pmc551_point (mtd, soff_hi, priv->asize, retlen, &ptr);
+			pmc551_point(mtd, soff_hi, priv->asize, retlen,
+				     (void **)&ptr, NULL);
                 }
                 memcpy(copyto, ptr, eoff_lo);
                 copyto += eoff_lo;
@@ -305,7 +313,7 @@ static int pmc551_write (struct mtd_info
         soff_lo = to & (priv->asize - 1);
         eoff_lo = end & (priv->asize - 1);
 
-	pmc551_point (mtd, to, len, retlen, &ptr);
+	pmc551_point(mtd, to, len, retlen, (void **)&ptr, NULL);
 
         if (soff_hi == eoff_hi) {
                 /* The whole thing fits within one access, so just one shot
@@ -325,7 +333,8 @@ static int pmc551_write (struct mtd_info
                                 goto out;
                         }
                         soff_hi += priv->asize;
-			pmc551_point (mtd, soff_hi, priv->asize, retlen, &ptr);
+			pmc551_point(mtd, soff_hi, priv->asize, retlen,
+				     (void **)&ptr, NULL);
                 }
                 memcpy(ptr, copyfrom, eoff_lo);
                 copyfrom += eoff_lo;
diff -r 907888e95d29 drivers/mtd/devices/slram.c
--- a/drivers/mtd/devices/slram.c	Tue Sep 02 14:27:46 2008 -0700
+++ b/drivers/mtd/devices/slram.c	Tue Sep 02 14:29:14 2008 -0700
@@ -76,8 +76,9 @@ static slram_mtd_list_t *slram_mtdlist =
 static slram_mtd_list_t *slram_mtdlist = NULL;
 
 static int slram_erase(struct mtd_info *, struct erase_info *);
-static int slram_point(struct mtd_info *, loff_t, size_t, size_t *, u_char **);
-static void slram_unpoint(struct mtd_info *, u_char *, loff_t,	size_t);
+static int slram_point(struct mtd_info *, loff_t, size_t, size_t *, void **,
+		unsigned long *);
+static void slram_unpoint(struct mtd_info *, loff_t, size_t);
 static int slram_read(struct mtd_info *, loff_t, size_t, size_t *, u_char *);
 static int slram_write(struct mtd_info *, loff_t, size_t, size_t *, const u_char *);
 
@@ -104,16 +105,20 @@ static int slram_erase(struct mtd_info *
 }
 
 static int slram_point(struct mtd_info *mtd, loff_t from, size_t len,
-		size_t *retlen, u_char **mtdbuf)
+		       size_t *retlen, void **virt, unsigned long *phys)
 {
 	slram_priv_t *priv = mtd->priv;
 
-	*mtdbuf = priv->start + from;
+	/* can we return a physical address with this driver? */
+	if (phys)
+		return -EINVAL;
+
+	*virt = priv->start + from;
 	*retlen = len;
 	return(0);
 }
 
-static void slram_unpoint(struct mtd_info *mtd, u_char *addr, loff_t from, size_t len)
+static void slram_unpoint(struct mtd_info *mtd, loff_t from, size_t len)
 {
 }
 
diff -r 907888e95d29 drivers/mtd/maps/uclinux.c
--- a/drivers/mtd/maps/uclinux.c	Tue Sep 02 14:27:46 2008 -0700
+++ b/drivers/mtd/maps/uclinux.c	Tue Sep 02 14:29:14 2008 -0700
@@ -45,10 +45,12 @@ struct mtd_partition uclinux_romfs[] = {
 /****************************************************************************/
 
 int uclinux_point(struct mtd_info *mtd, loff_t from, size_t len,
-	size_t *retlen, u_char **mtdbuf)
+	size_t *retlen, void **virt, unsigned long *phys)
 {
 	struct map_info *map = mtd->priv;
-	*mtdbuf = (u_char *) (map->virt + ((int) from));
+	*virt = map->virt + from;
+	if (phys)
+		*phys = map->phys + from;
 	*retlen = len;
 	return(0);
 }
diff -r 907888e95d29 drivers/mtd/mtdpart.c
--- a/drivers/mtd/mtdpart.c	Tue Sep 02 14:27:46 2008 -0700
+++ b/drivers/mtd/mtdpart.c	Tue Sep 02 14:29:14 2008 -0700
@@ -64,7 +64,7 @@ static int part_read (struct mtd_info *m
 }
 
 static int part_point (struct mtd_info *mtd, loff_t from, size_t len, 
-			size_t *retlen, u_char **buf)
+			size_t *retlen, void **virt, unsigned long *phys)
 {
 	struct mtd_part *part = PART(mtd);
 	if (from >= mtd->size)
@@ -72,13 +72,13 @@ static int part_point (struct mtd_info *
 	else if (from + len > mtd->size)
 		len = mtd->size - from;
 	return part->master->point (part->master, from + part->offset, 
-				    len, retlen, buf);
+				    len, retlen, virt, phys);
 }
-static void part_unpoint (struct mtd_info *mtd, u_char *addr, loff_t from, size_t len)
+static void part_unpoint (struct mtd_info *mtd, loff_t from, size_t len)
 {
 	struct mtd_part *part = PART(mtd);
 
-	part->master->unpoint (part->master, addr, from + part->offset, len);
+	part->master->unpoint (part->master, from + part->offset, len);
 }
 
 
diff -r 907888e95d29 fs/jffs2/nodelist.c
--- a/fs/jffs2/nodelist.c	Tue Sep 02 14:27:46 2008 -0700
+++ b/fs/jffs2/nodelist.c	Tue Sep 02 14:29:14 2008 -0700
@@ -289,10 +289,10 @@ int jffs2_get_inode_nodes(struct jffs2_s
 #ifndef __ECOS
 					if (c->mtd->point) {
 						err = c->mtd->point (c->mtd, ref_offset(ref) + sizeof(node.i), je32_to_cpu(node.i.csize),
-								     &retlen, &buf);
+								     &retlen, (void*)&buf, NULL);
 						if (!err && retlen < je32_to_cpu(node.i.csize)) {
 							D1(printk(KERN_DEBUG "MTD point returned len too short: 0x%zx\n", retlen));
-							c->mtd->unpoint(c->mtd, buf, ref_offset(ref) + sizeof(node.i), je32_to_cpu(node.i.csize));
+							c->mtd->unpoint(c->mtd, ref_offset(ref) + sizeof(node.i), je32_to_cpu(node.i.csize));
 						} else if (err){
 							D1(printk(KERN_DEBUG "MTD point failed %d\n", err));
 						} else
@@ -318,7 +318,7 @@ int jffs2_get_inode_nodes(struct jffs2_s
 						kfree(buf);
 #ifndef __ECOS
 					else
-						c->mtd->unpoint(c->mtd, buf, ref_offset(ref) + sizeof(node.i), je32_to_cpu(node.i.csize));
+						c->mtd->unpoint(c->mtd, ref_offset(ref) + sizeof(node.i), je32_to_cpu(node.i.csize));
 #endif
 
 					if (crc != je32_to_cpu(node.i.data_crc)) {
diff -r 907888e95d29 fs/jffs2/scan.c
--- a/fs/jffs2/scan.c	Tue Sep 02 14:27:46 2008 -0700
+++ b/fs/jffs2/scan.c	Tue Sep 02 14:29:14 2008 -0700
@@ -85,11 +85,12 @@ int jffs2_scan_medium(struct jffs2_sb_in
 	size_t pointlen;
 
 	if (c->mtd->point) {
-		ret = c->mtd->point (c->mtd, 0, c->mtd->size, &pointlen, &flashbuf);
+		ret = c->mtd->point (c->mtd, 0, c->mtd->size, &pointlen,
+				     (void **)&flashbuf, NULL);
 		if (!ret && pointlen < c->mtd->size) {
 			/* Don't muck about if it won't let us point to the whole flash */
 			D1(printk(KERN_DEBUG "MTD point returned len too short: 0x%zx\n", pointlen));
-			c->mtd->unpoint(c->mtd, flashbuf, 0, c->mtd->size);
+			c->mtd->unpoint(c->mtd, 0, c->mtd->size);
 			flashbuf = NULL;
 		}
 		if (ret)
@@ -252,7 +253,7 @@ int jffs2_scan_medium(struct jffs2_sb_in
 		kfree(flashbuf);
 #ifndef __ECOS
 	else 
-		c->mtd->unpoint(c->mtd, flashbuf, 0, c->mtd->size);
+		c->mtd->unpoint(c->mtd, 0, c->mtd->size);
 #endif
 	return ret;
 }
diff -r 907888e95d29 include/linux/mtd/mtd.h
--- a/include/linux/mtd/mtd.h	Tue Sep 02 14:27:46 2008 -0700
+++ b/include/linux/mtd/mtd.h	Tue Sep 02 14:29:14 2008 -0700
@@ -93,10 +93,12 @@ struct mtd_info {
 	int (*erase) (struct mtd_info *mtd, struct erase_info *instr);
 
 	/* This stuff for eXecute-In-Place */
-	int (*point) (struct mtd_info *mtd, loff_t from, size_t len, size_t *retlen, u_char **mtdbuf);
+	/* phys is optional and may be set to NULL */
+	int (*point) (struct mtd_info *mtd, loff_t from, size_t len,
+			size_t *retlen, void **virt, unsigned long *phys);
 
 	/* We probably shouldn't allow XIP if the unpoint isn't a NULL */
-	void (*unpoint) (struct mtd_info *mtd, u_char * addr, loff_t from, size_t len);
+	void (*unpoint) (struct mtd_info *mtd, loff_t from, size_t len);
 
 
 	int (*read) (struct mtd_info *mtd, loff_t from, size_t len, size_t *retlen, u_char *buf);
diff -r 907888e95d29 include/linux/mtd/pmc551.h
--- a/include/linux/mtd/pmc551.h	Tue Sep 02 14:27:46 2008 -0700
+++ b/include/linux/mtd/pmc551.h	Tue Sep 02 14:29:14 2008 -0700
@@ -36,8 +36,9 @@ struct mypriv {
  * Function Prototypes
  */
 static int pmc551_erase(struct mtd_info *, struct erase_info *);
-static void pmc551_unpoint(struct mtd_info *, u_char *, loff_t, size_t);
-static int pmc551_point (struct mtd_info *mtd, loff_t from, size_t len, size_t *retlen, u_char **mtdbuf);
+static void pmc551_unpoint(struct mtd_info *, loff_t, size_t);
+static int pmc551_point(struct mtd_info *mtd, loff_t from, size_t len,
+		size_t *retlen, void **virt, unsigned long *phys);
 static int pmc551_read(struct mtd_info *, loff_t, size_t, size_t *, u_char *);
 static int pmc551_write(struct mtd_info *, loff_t, size_t, size_t *, const u_char *);        
 
