diff -r 81e43de93a44 arch/um/include/sysdep-i386/stub.h
--- a/arch/um/include/sysdep-i386/stub.h	Tue Sep 16 17:15:46 2008 -0700
+++ b/arch/um/include/sysdep-i386/stub.h	Fri Nov 21 01:45:37 2008 -0800
@@ -7,9 +7,9 @@
 #define __SYSDEP_STUB_H
 
 #include <sys/mman.h>
+#include <sys/user.h>
 #include <asm/ptrace.h>
 #include <asm/unistd.h>
-#include <asm/page.h>
 #include "stub-data.h"
 #include "kern_constants.h"
 #include "uml-config.h"
@@ -17,6 +17,7 @@
 extern void stub_segv_handler(int sig);
 extern void stub_clone_handler(void);
 
+#define PAGE_SHIFT 12
 #define STUB_SYSCALL_RET EAX
 #define STUB_MMAP_NR __NR_mmap2
 #define MMAP_OFFSET(o) ((o) >> PAGE_SHIFT)
diff -r 81e43de93a44 arch/um/kernel/skas/clone.c
--- a/arch/um/kernel/skas/clone.c	Tue Sep 16 17:15:46 2008 -0700
+++ b/arch/um/kernel/skas/clone.c	Fri Nov 21 01:45:37 2008 -0800
@@ -3,7 +3,7 @@
 #include <sys/mman.h>
 #include <sys/time.h>
 #include <asm/unistd.h>
-#include <asm/page.h>
+#include <sys/user.h>
 #include "ptrace_user.h"
 #include "skas.h"
 #include "stub-data.h"
diff -r 81e43de93a44 arch/um/os-Linux/aio.c
--- a/arch/um/os-Linux/aio.c	Tue Sep 16 17:15:46 2008 -0700
+++ b/arch/um/os-Linux/aio.c	Fri Nov 21 01:45:37 2008 -0800
@@ -76,8 +76,7 @@
 				.aio_nbytes	= len,
 				.aio_offset	= offset,
 				.aio_reserved1	= 0,
-				.aio_reserved2	= 0,
-				.aio_reserved3	= 0 });
+				.aio_reserved2	= 0 });
 
 	switch(type){
 	case AIO_READ:
diff -r 81e43de93a44 arch/um/os-Linux/main.c
--- a/arch/um/os-Linux/main.c	Tue Sep 16 17:15:46 2008 -0700
+++ b/arch/um/os-Linux/main.c	Fri Nov 21 01:45:37 2008 -0800
@@ -12,7 +12,6 @@
 #include <sys/resource.h>
 #include <sys/mman.h>
 #include <sys/user.h>
-#include <asm/page.h>
 #include "user_util.h"
 #include "kern_util.h"
 #include "mem_user.h"
diff -r 81e43de93a44 arch/um/os-Linux/skas/mem.c
--- a/arch/um/os-Linux/skas/mem.c	Tue Sep 16 17:15:46 2008 -0700
+++ b/arch/um/os-Linux/skas/mem.c	Fri Nov 21 01:45:37 2008 -0800
@@ -8,7 +8,7 @@
 #include <string.h>
 #include <sys/mman.h>
 #include <sys/wait.h>
-#include <asm/page.h>
+#include <sys/user.h>
 #include <asm/unistd.h>
 #include "mem_user.h"
 #include "mem.h"
diff -r 81e43de93a44 arch/um/os-Linux/start_up.c
--- a/arch/um/os-Linux/start_up.c	Tue Sep 16 17:15:46 2008 -0700
+++ b/arch/um/os-Linux/start_up.c	Fri Nov 21 01:45:37 2008 -0800
@@ -18,7 +18,7 @@
 #include <sys/wait.h>
 #include <sys/mman.h>
 #include <asm/unistd.h>
-#include <asm/page.h>
+#include <sys/user.h>
 #include <sys/types.h>
 #include "user_util.h"
 #include "kern_util.h"
diff -r 81e43de93a44 arch/um/sys-i386/ptrace_user.c
--- a/arch/um/sys-i386/ptrace_user.c	Tue Sep 16 17:15:46 2008 -0700
+++ b/arch/um/sys-i386/ptrace_user.c	Fri Nov 21 01:45:37 2008 -0800
@@ -9,7 +9,7 @@
 #include <unistd.h>
 #include "ptrace_user.h"
 /* Grr, asm/user.h includes asm/ptrace.h, so has to follow ptrace_user.h */
-#include <asm/user.h>
+#include <sys/user.h>
 #include "kern_util.h"
 #include "sysdep/thread.h"
 #include "user.h"
diff -r 81e43de93a44 arch/um/sys-i386/user-offsets.c
--- a/arch/um/sys-i386/user-offsets.c	Tue Sep 16 17:15:46 2008 -0700
+++ b/arch/um/sys-i386/user-offsets.c	Fri Nov 21 01:45:37 2008 -0800
@@ -1,7 +1,7 @@
 #include <stdio.h>
 #include <signal.h>
 #include <asm/ptrace.h>
-#include <asm/user.h>
+#include <sys/user.h>
 #include <stddef.h>
 #include <sys/poll.h>
 
@@ -48,8 +48,8 @@
 	OFFSET(HOST_SC_FXSR_ENV, _fpstate, _fxsr_env);
 
 	DEFINE(HOST_FRAME_SIZE, FRAME_SIZE);
-	DEFINE_LONGS(HOST_FP_SIZE, sizeof(struct user_i387_struct));
-	DEFINE_LONGS(HOST_XFP_SIZE, sizeof(struct user_fxsr_struct));
+	DEFINE_LONGS(HOST_FP_SIZE, sizeof(struct user_fpregs_struct));
+	DEFINE_LONGS(HOST_XFP_SIZE, sizeof(struct user_fpxregs_struct));
 
 	DEFINE(HOST_IP, EIP);
 	DEFINE(HOST_SP, UESP);
diff -r 81e43de93a44 scripts/mod/sumversion.c
--- a/scripts/mod/sumversion.c	Tue Sep 16 17:15:46 2008 -0700
+++ b/scripts/mod/sumversion.c	Fri Nov 21 01:45:37 2008 -0800
@@ -7,6 +7,7 @@
 #include <ctype.h>
 #include <errno.h>
 #include <string.h>
+#include <limits.h>
 #include "modpost.h"
 
 /*
