diff --git a/kernel/mutex.c b/kernel/mutex.c
index 8c71cf7..dd2bf55 100644
--- a/kernel/mutex.c
+++ b/kernel/mutex.c
@@ -93,7 +93,7 @@ void inline fastcall __sched mutex_lock(struct mutex *lock)
 
 EXPORT_SYMBOL(mutex_lock);
 
-static void fastcall noinline __sched
+static __used void fastcall noinline __sched
 __mutex_unlock_slowpath(atomic_t *lock_count);
 
 /***
@@ -189,7 +189,7 @@ __mutex_lock_common(struct mutex *lock, long state, unsigned int subclass)
 	return 0;
 }
 
-static void fastcall noinline __sched
+static void __used fastcall noinline __sched
 __mutex_lock_slowpath(atomic_t *lock_count)
 {
 	struct mutex *lock = container_of(lock_count, struct mutex, count);
