diff -r b11e3ede7d80 -r 09dc3aa912ac Makefile
--- a/Makefile	Tue Jul 10 23:10:09 2007 -0700
+++ b/Makefile	Tue Jul 10 23:42:13 2007 -0700
@@ -477,19 +477,26 @@ endif # $(dot-config)
 # Defaults vmlinux but it is usually overridden in the arch makefile
 all: vmlinux
 
-ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
-CFLAGS		+= -Os
-else
-CFLAGS		+= -O2
-endif
+#ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
+#CFLAGS		+= -O0
+#else
+CFLAGS		+= -O1
+CFLAGS		+= -funroll-loops
+CFLAGS		+= -fno-defer-pop
+CFLAGS		+= -fno-thread-jumps
+CFLAGS		+= -fno-branch-probabilities 
+CFLAGS		+= -fno-cprop-registers
+CFLAGS		+= -fno-guess-branch-probability
+CFLAGS		+= -fno-omit-frame-pointer
+#endif
 
 include $(srctree)/arch/$(ARCH)/Makefile
 
-ifdef CONFIG_FRAME_POINTER
-CFLAGS		+= -fno-omit-frame-pointer $(call cc-option,-fno-optimize-sibling-calls,)
-else
-CFLAGS		+= -fomit-frame-pointer
-endif
+#ifdef CONFIG_FRAME_POINTER
+#CFLAGS		+= -fno-omit-frame-pointer $(call cc-option,-fno-optimize-sibling-calls,)
+#else
+#CFLAGS		+= -fomit-frame-pointer
+#endif
 
 ifdef CONFIG_UNWIND_INFO
 CFLAGS		+= -fasynchronous-unwind-tables
